### PR TITLE
Added type definition for `codegen.macro`

### PR DESCRIPTION
**Pull request by charlee**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/codegen.macro/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+// Type definitions for codegen.macro 3.0
+// Project: https://github.com/kentcdodds/codegen.macro
+// Definitions by: Charlee Li <https://github.com/charlee>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+declare function codegen(strings: TemplateStringsArray, ...values: any[]): any;
```
Comment (uniqueiniquity) [inline types/codegen.macro/index.d.ts]: I would suggest the return type be `unknown`. This way, consumers of the library will have to cast the result of `codegen` in order to use it. As you have it now, you can "cast" it using type annotations as you do in your tests, but if they don't do this, then there's no error and they might introduce `any` into their program.

**Review by uniqueiniquity — CHANGES_REQUESTED** [review]

File: types/codegen.macro/index.d.ts (hunk 1 above)
Comment (charlee) [reply types/codegen.macro/index.d.ts]: Yes totally agree. But do we still need to support TypeScript versions older than 3.0?

Comment (uniqueiniquity) [reply types/codegen.macro/index.d.ts]: Ah, good point, I sometimes forget how new `unknown` is. In that case, you could use `{} | null | undefined`.

Comment (charlee) [reply types/codegen.macro/index.d.ts]: I think the return value of `codegen` can be literally anything, not limited to `{}` or `null` or `undefined`.
I feel `any` should be a good representation of its return type, or `unknown` to be more accurate, but any specific types could limit its potential.

Comment (uniqueiniquity) [reply types/codegen.macro/index.d.ts]: `unknown` and `{} | null | undefined` are actually the same thing. But I understand your concern, so I'm fine merging either that or `any`; whichever you prefer.

Comment (charlee) [reply types/codegen.macro/index.d.ts]: Thanks for pointing that out. I tried `{} | null | undefined` and even `{} | null | undefined | number | string`, not sure if I was doing something wrong, but looks like it was not compatible with the types given in the test case. For example, `{}` is not assignable to type `{ [key: string]: string; }`.
Given that, I would like it to be `any` which makes it much simpler.

**Comment by charlee** [comment]
Hi @uniqueiniquity , as commented above I would like to choose using `any`. Do I need to do anything else?

**Review by uniqueiniquity — APPROVED** [review]

**Comment by uniqueiniquity** [comment]
Congratulations on your first DefinitelyTyped contribution!
Thank you for being a part of the community!

**Comment by uniqueiniquity** [comment]
Sorry for the delay!

**Pull request merged**